### PR TITLE
Shallow git fetch before mlflow run

### PR DESCRIPTION
**Pull request by marijncv**
Signed-off-by: Marijn Valk <marijncv@hotmail.com>

## What changes are proposed in this pull request?

This PR updates the project run utils such that it fetches the user code repository with depth=1. This is a small optimization to reduce the size of the .git folder in the repository that is cloned (and subsequently sent to dbfs when running on Databricks backend). The revisions are not needed when running a project.

## How is this patch tested?

Used the `mlflow run` command with a git repository uri. Tested with and without version specified.

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/projects/test_utils.py [hunk 1]
```diff
@@ -92,6 +96,22 @@ def test__fetch_project(local_git_repo, local_git_repo_uri, zipped_repo, httpser
...
+        (GIT_PROJECT_BRANCH, GIT_PROJECT_BRANCH, does_not_raise()),
+        ("non-version", "", pytest.raises(ExecutionException)),
+    ],
+)
+def test__fetch_git_repo(local_git_repo, local_git_repo_uri, version, expected_version, raises):
+    # Verify that the correct branch is checked out
+    with raises:
+        _fetch_git_repo(local_git_repo_uri, version, local_git_repo)
+        repo = git.Repo(local_git_repo)
+        assert repo.active_branch.name == expected_version
```
Comment (harupy) [inline tests/projects/test_utils.py:112]: Can we split this test like below?

```python
@pytest.mark.parametrize(
    "version,expected_version", [(None, "master"), (GIT_PROJECT_BRANCH, GIT_PROJECT_BRANCH)],
)
def test__fetch_git_repo(local_git_repo, local_git_repo_uri, version, expected_version):
    # Verify that the correct branch is checked out
    _fetch_git_repo(local_git_repo_uri, version, local_git_repo)
    repo = git.Repo(local_git_repo)
    assert repo.active_branch.name == expected_version


def test_fetching_non_existing_version_fails(local_git_repo, local_git_repo_uri):
    with pytest.raises(ExecutionException, match="Unable to checkout"):
        _fetch_git_repo(local_git_repo_uri, "non-version", local_git_repo)
```

Comment (marijncv) [reply tests/projects/test_utils.py]: Hey @harupy, thanks for the pointer. The test is split up now

**Comment by harupy** [comment]
@marijncv I locally tested this PR using [our example repo](https://github.com/mlflow/mlflow-example.git) and it worked!

**Review by harupy — APPROVED** [review]
LGTM

**Comment by harupy** [comment]
@marijncv Can you fix the lint error?

**Comment by marijncv** [comment]
@harupy, I removed the unused import. Thanks for the review!

**Comment by adeandrade** [comment]
@harupy @marijncv This PR is causing issues. We cannot run anything but the latest commit. Can we revert or extend `GIT_FETCH_DEPTH` to something reasonable?

**Pull request merged**